packages/event-bus-redis/src/services/event-bus-redis.ts

```diff
@@ -0,0 +1,210 @@
+import { ConfigModule, Logger } from "@medusajs/medusa"
+import { InternalModuleDeclaration } from "@medusajs/modules-sdk"
+import { EventBusTypes } from "@medusajs/types"
+import { EventBusUtils } from "@medusajs/utils"
+import { BulkJobOptions, JobsOptions, Queue, Worker } from "bullmq"
+import { Redis } from "ioredis"
+import { BullJob, EmitOptions, EventBusRedisModuleOptions } from "../types"
+
+type InjectedDependencies = {
+  logger: Logger
+  configModule: ConfigModule
+  eventBusRedisConnection: Redis
+}
+
+/**
+ * Can keep track of multiple subscribers to different events and run the
+ * subscribers when events happen. Events will run asynchronously.
+ */
+export default class RedisEventBusService extends EventBusUtils.AbstractEventBusModuleService {
+  protected readonly config_: ConfigModule
+  protected readonly logger_: Logger
+  protected readonly moduleOptions_: EventBusRedisModuleOptions
+  protected readonly moduleDeclaration_: InternalModuleDeclaration
+
+  protected queue_: Queue
+
+  constructor(
+    { configModule, logger, eventBusRedisConnection }: InjectedDependencies,
+    moduleOptions: EventBusRedisModuleOptions = {},
+    moduleDeclaration: InternalModuleDeclaration
+  ) {
+    // @ts-ignore
+    super(...arguments)
+
+    this.moduleOptions_ = moduleOptions
+    this.config_ = configModule
+    this.logger_ = logger
+
+    this.queue_ = new Queue(moduleOptions.queueName ?? `events-queue`, {
+      prefix: `${this.constructor.name}`,
+      ...(moduleOptions.queueOptions ?? {}),
+      connection: eventBusRedisConnection,
+    })
+
+    // Register our worker to handle emit calls
+    new Worker(moduleOptions.queueName ?? "events-queue", this.worker_, {
+      prefix: `${this.constructor.name}`,
+      ...(moduleOptions.workerOptions ?? {}),
+      connection: eventBusRedisConnection,
+    })
+  }
+
+  /**
+   * Emit a single event
+   * @param {string} eventName - the name of the event to be process.
+   * @param data - the data to send to the subscriber.
+   * @param options - options to add the job with
+   */
+  async emit<T>(
+    eventName: string,
+    data: T,
+    options: Record<string, unknown>
+  ): Promise<void>
+
+  /**
+   * Emit a number of events
+   * @param {EmitData} data - the data to send to the subscriber.
+   */
+  async emit<T>(data: EventBusTypes.EmitData<T>[]): Promise<void>
+
+  async emit<T, TInput extends string | EventBusTypes.EmitData<T>[] = string>(
+    eventNameOrData: TInput,
+    data?: T,
+    options: BulkJobOptions | JobsOptions = {}
+  ): Promise<void> {
+    const globalJobOptions = this.moduleOptions_.jobOptions ?? {}
+
+    const isBulkEmit = Array.isArray(eventNameOrData)
+
+    const opts = {
+      // default options
+      removeOnComplete: true,
+      attempts: 1,
+      // global options
+      ...globalJobOptions,
+    } as EmitOptions
+
+    const events = isBulkEmit
+      ? eventNameOrData.map((event) => ({
+          name: event.eventName,
```

Comment:
question: Is that intentional? I am fine with it but otherwise it is set to a default name internally

Reply:
Re raising the question

Reply:
`name` is required in `bullmq`, so omitting it will make Typescript scream at me :)

Reply:
Got it 🙏